src/Interpreters/ExpressionActions.cpp

```diff
@@ -68,25 +74,182 @@ ExpressionActionsPtr ExpressionActions::clone() const
+
+    for (const auto & node : nodes)
+    {
+        if (node.type == ActionsDAG::ActionType::FUNCTION && node.function_base->isShortCircuit(short_circuit_settings, node.children.size()) && !node.children.empty())
+        {
+            short_circuit_nodes.insert(&node);
+            std::deque<const ActionsDAG::Node *> queue;
+
+            /// For some short-circuit function we shouldn't enable lazy execution for actions that are common
+            /// descendants of different function arguments. Example: if(cond, expr1(..., expr, ...), expr2(..., expr, ...))).
+            /// For each node we will store the index of argument that is it's ancestor. If node has two
+            /// parents with different argument ancestor, this node is common descendants of two different function arguments.
+            std::unordered_map<const ActionsDAG::Node *, int> argument_ancestor;
+
+            size_t i = 0;
+            if (!short_circuit_settings.enable_lazy_execution_for_first_argument)
+            {
+                ++i;
+                skip_nodes[&node].insert(node.children[0]);
+            }
+
+            for (; i < node.children.size(); ++i)
+            {
+                queue.push_back(node.children[i]);
+                if (!short_circuit_settings.enable_lazy_execution_for_common_descendants_of_arguments)
+                {
+                    /// Check repeated argument.
+                    if (argument_ancestor.contains(node.children[i]))
+                        skip_nodes[&node].insert(node.children[i]);
+                    else
+                        argument_ancestor[node.children[i]] = i;
+                }
+            }
+
+            while (!queue.empty())
+            {
+                const ActionsDAG::Node * cur = queue.front();
+                queue.pop_front();
 
+                if (skip_nodes[&node].contains(cur))
+                    continue;
+
+                bool is_used_only_in_short_circuit_functions = true;
+                /// If action is used in result, we can't enable lazy execution.
+                if (data[reverse_index.at(cur)].used_in_result)
+                    is_used_only_in_short_circuit_functions = false;
+                else
+                {
+                    for (const auto * parent : data[reverse_index.at(cur)].parents)
+                    {
+                        /// Mark this node as needed outside shirt-circuit functions if (it's parent is not
+                        /// short-circuit or node in it's skip list) AND parent is marked as needed outside.
+                        if ((!short_circuit_nodes.contains(parent) || skip_nodes[parent].contains(cur)) && (!used_only_in_short_circuit_functions.contains(parent) || !used_only_in_short_circuit_functions[parent]))
+                        {
+                            is_used_only_in_short_circuit_functions = false;
+                            break;
+                        }
```

Comment:
It is probably nothing wrong with this code, but still difficult to understand.
I suspect that a query `select if(number > 0, intDiv(42, number), 0), if(number != 0, intDiv(42, number), 0) from numbers(4)` now works only because we mark `intDiv(42, number)`  as `used_only_in_short_circuit_functions` in the **second** if.

I don't feel like this code is strong solid...

Reply:
I have rewritten the code in ExpressionActions with another approach, please take a look at it.